Vars to set navbar text color and hover-over element color

jennydaman opened the pull request:
Dark navbars look pretty slick (just look at Github!)

All the changes mentioned below affect `_config.yaml` and `css/main.css`:

1. `navbar-col` applies to submenus in the navbar (`.navlinks-children`) for consistency. 
2. `navbar-text-col` determines the color of text in the navbar and its children. This should be set to increase contrast with a dark `navbar-col` color. 
3. `hover-col` will set the color that some elements change to when hovered over. Links in the navbar,
 links to blog posts, and the social media buttons in the footer are some elements are colored by `hover-col`. (This does not affect links to external sites.)

![screenshot from 2017-09-30 23-21-28](https://user-images.githubusercontent.com/20404439/31051433-1da503ce-a636-11e7-9172-b01e23f86e2e.png)

daattali commented:
Thank you for the great PR that even includes description and screenshot and even a changelog update. You've clearly done this before, thank you! That saves me a lot of time.

Can you change `#EEE` to `#EEEEEE`? And for the hover colour, it looks like you added a rule to the plain `a:hover { ... }`, are you sure that won't affect most other links on the page?

jennydaman commented:
I'm honestly flattered by your words, this is my first PR...

Yesterday's headache didn't play nice with my senses, so you are right: normal links to external sites will change color on hover. Not going to try to explain what I was thinking. Anyway, talk is cheap. I'll show you the code. 

I changed the commit message in 558bcf6df611e5003001ab8bb77fa38fb3761354 with a force push to avoid confusion about URL hover color. In b501488256ba2279c6c90c8e822be78452122611, I fixed a visual bug that I introduced in 558bcf6df611e5003001ab8bb77fa38fb3761354 by accident which had caused the navbar to resize when the `navlinks-children` submenu showed. Additionally, `#EEE` was corrected to `#EEEEEE`

daattali commented:
Usually a person's first PR is a mess and undocumented and unexplained and is not fun to deal with. Yours is great :) I'll test your changes hopefully within a week and then merge. Thanks!

daattali commented on css/main.css:
```diff
@@ -612,6 +638,10 @@ footer .theme-by {
   border: 1px solid #0085a1;
+  {% if site.hover-col %}
+  background: {{ site.hover-col }};
+  border: {{ site.hover-col }};
```
@jennydaman shouldn't this be `border-color` rather than just `border`?

jennydaman commented:
Was careless, fixed.

I also pulled from upstream to revolve the merge conflicts.

daattali commented:
I changed the changelog because you didn't include your name! Could you remove the changes made to the gem lockfile since they're not part of this feature?

daattali commented:
@jennydaman Did you also test the new config options on small screens (make your window narrow) to make sure the responsive menu also looks fine with these changes?

jennydaman commented:
Everything is a disaster. 

So, I restored the `Gemfile.lock` from master, so that it includes only the configuration for mingw (I'm on Linux with native Ruby gems, so my machine keeps writing over that file. I'll ignore these changes locally.) 

I didn't touch the color of the "navbar-toggle" on low-width screens because that was set in `bootstrap.css`. I revisited this issue because you mentioned it, and it led me to a bug on master...

What I'm referring to when I mention `navbar-toggle` is this button that appears on low-width screens. It expands and displays the `navlinks-children` when clicked on.
![screenshot from 2017-10-16 21-28-33](https://user-images.githubusercontent.com/20404439/31642452-0954382e-b2b9-11e7-8ec2-58d693831d5e.png)

## The issue

`bootstrap.min.css` sets the darker grey color to `navbar-toggle` using the `:focus` state. The issue is, the `:focus` state persists after the navbar's children are hidden. What this means is that `navbar-toggle` stays grey after being clicked on until you click somewhere else. 

### Steps to reproduce

1. Set screen width to less than 760px so that `navbar-toggle` appears.
2. Click `navbar-toggle` to give it the `:focus` state. It will turn a darker shade.
3. Click `navbar-toggle` again to hide the `navlinks-children`. 
4. Even if you move your mouse off the button, the button retains its `:focus` state, and the button stays the darker shade of grey. 

*****

To fix this issue, I reacted a new branch in my fork: [nbtoggle-color](https://github.com/jennydaman/beautiful-jekyll/tree/nbtoggle-color). This is a lazy solution just to prove the expected behavior. I changed `_layouts/base.html` to use the uncompressed `css/bootstrap.css` for convenience. 

On line 4420 in `css/bootstrap.css`, I changed `.navbar-toggle:focus` to `.navbar-toggle:active, button[aria-expanded="true"]`

If you agree with this change, I can directly patch `css/bootstrap.min.css` instead and open a new pull request. 

Making the `navbar-toggle` color consistent with my original feature proposal of link hover colors will be rather inconvenient because the colors are hard-coded into `bootstrap.css`. Definitely possible though.

daattali commented:
Thanks for taking a look and acting on it. Looks like this very simple PR
is going down quite a rabbit hole unfortunately!
I haven't noticed that before, but I would call that a bug in bootstrap.
I'm really uncomfortable with the idea of changing the bootstrap css, any
3rd party scripts should be left untouched so that it's easier to upgrade
and to isolate issues. Perhaps overwriting the :focus behaviour in our css
can help?

---
Dean Attali
President & CEO
AttaliTech Ltd
http://AttaliTech.com <http://attalitech.com>

On 16 October 2017 at 21:41, Jennings Zhang <notifications@github.com>
wrote:

> Everything is a disaster.
>
> So, I restored the Gemfile.lock from master, so that it includes only the
> configuration for mingw (I'm on Linux with native Ruby gems, so my machine
> keeps writing over that file. I'll ignore these changes locally.)
>
> I didn't touch the color of the "navbar-toggle" on low-width screens
> because that was set in bootstrap.css. I revisited this issue because you
> mentioned it, and it led me to a bug on master...
>
> What I'm referring to when I mention navbar-toggle is this button that
> appears on low-width screens. It expands and displays the
> navlinks-children when clicked on.
> [image: screenshot from 2017-10-16 21-28-33]
> <https://user-images.githubusercontent.com/20404439/31642452-0954382e-b2b9-11e7-8ec2-58d693831d5e.png>
> The issue
>
> bootstrap.min.css sets the darker grey color to navbar-toggle using the
> :focus state. The issue is, the :focus state persists after the navbar's
> children are hidden. What this means is that navbar-toggle stays grey
> after being clicked on until you click somewhere else.
> Steps to reproduce
>
>    1. Set screen width to less than 760px so that navbar-toggle appears.
>    2. Click navbar-toggle to give it the :focus state. It will turn a
>    darker shade.
>    3. Click navbar-toggle again to hide the navlinks-children.
>    4. Even if you move your mouse off the button, the button retains its
>    :focus state, and the button stays the darker shade of grey.
>
> ------------------------------
>
> To fix this issue, I reacted a new branch in my fork: nbtoggle-color
> <https://github.com/jennydaman/beautiful-jekyll/tree/nbtoggle-color>.
> This is a lazy solution just to prove the expected behavior. I changed
> _layouts/base.html to use the uncompressed css/bootstrap.css for
> convenience.
>
> On line 4420 in css/bootstrap.css, I changed .navbar-toggle:focus to .navbar-toggle:active,
> button[aria-expanded="true"]
>
> If you agree with this change, I can directly patch css/bootstrap.min.css
> instead and open a new pull request.
>
> Making the navbar-toggle color consistent with my original feature
> proposal of link hover colors will be rather inconvenient because the
> colors are hard-coded into bootstrap.css. Definitely possible though.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/daattali/beautiful-jekyll/pull/256#issuecomment-337092241>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6IFKxUP0GdVammC-YKJnOdDPq22T-eks5stAXlgaJpZM4PpxPv>
> .
>

daattali commented:
@jennydaman is this PR completely fixed and ready for review/merge?

jennydaman commented:
So the colors work fine, it's just that setting the color of `.navbar-toggle` is pretty sketchy because of #268. Additionally, issue #268 becomes more obvious and pronounced if the colors are changed to some wacky colors...

To be clear, the `:navbar-toggle` bug is present on master. This PR doesn't add any bugs.. It's just the concept of adding a "changing colors" feature is distasteful when bugs related to color are present on master.

Do you want to merge it anyway? If so, I guess there's a few more quick things to add, like options to change text color to so that it contrasts with a custom background color. (I'd also want to change the comments that suggest an orange/blue/pink color scheme, it's atrocious...) If you want to wait until #268 is resolved, then I don't want to waste my time.

daattali commented:
I agree with what you said in the issue - it'd be ok to just remove the hover effect on the toggle button. Since the toggle button only showed up on small screens, it's mostly used on touchscreen devices, where it's buggy. Toggle button on non-touchscreen means you have a very tiny browser, and you won't have a hover effect on it, that's fine. Do you want to do that?

jennydaman commented:
starting over...
